packages/components/src/segmented-control/segmented-control.tsx

```diff
@@ -0,0 +1,293 @@
+      ]),
+    })
+
+    return (
+      <button
+        ref={ref}
+        onClick={onClick}
+        className={segmentStyles({
+          active,
+          variant: children ? 'icon' : 'icon-only',
+          size,
+          type,
+        })}
+        value={value}
+      >
+        {iconWithColor}
+        {children}
+      </button>
+    )
+  },
+)
+
+IconButton.displayName = 'IconButton'
+
+const EmojiButton = forwardRef<
+  HTMLButtonElement,
+  SegmentButtonProps & {
+    emoji: string
+    value?: string | number
+    type?: 'grey' | 'dark-grey'
+  }
+>(
+  (
+    { children, emoji, onClick, active, size = '32', value, type = 'grey' },
+    ref,
+  ) => (
+    <button
+      ref={ref}
+      onClick={onClick}
+      className={segmentStyles({
+        active,
+        variant: children ? 'emoji' : 'emoji-only',
+        size,
+        type,
+      })}
+      value={value}
+    >
+      {emoji} {children}
+    </button>
+  ),
+)
+
+EmojiButton.displayName = 'EmojiButton'
+
+type Props<T> = {
+  children: ReactNode
+  activeSegment: T
+  onSegmentChange: (value: T) => void
+  size?: '24' | '32'
+  type?: 'grey' | 'dark-grey'
+}
+
+const SegmentedControl = <T extends string | number>(props: Props<T>) => {
+  const {
+    children,
+    activeSegment,
+    onSegmentChange,
+    size = '32',
+    type = 'grey',
+  } = props
+
+  const [indicatorStyle, setIndicatorStyle] = useState<React.CSSProperties>({})
+  const segmentRefs = useRef<(HTMLButtonElement | null)[]>([])
+
+  useEffect(() => {
+    const activeSegmentIndex = Children.toArray(children).findIndex(
+      child =>
+        isValidElement(child) &&
+        (child.props.value
+          ? child.props.value === activeSegment
+          : child.props.children === activeSegment),
+    )
+
+    const activeSegmentRef = segmentRefs.current[activeSegmentIndex]
+    if (activeSegmentRef) {
+      setIndicatorStyle({
+        width: `${activeSegmentRef.offsetWidth}px`,
+        transform: `translateX(${activeSegmentRef.offsetLeft}px)`,
+      })
+    }
+  }, [activeSegment, children])
+
+  const clonedChildren = Children.map(
+    children,
+    (child, index) =>
+      isValidElement(child) &&
+      cloneElement(child, {
+        ...child.props,
+        onClick: () =>
+          onSegmentChange(child.props.value || child.props.children),
+        active: child.props.value
+          ? child.props.value === activeSegment
+          : child.props.children === activeSegment,
+        ref: (el: HTMLButtonElement) => (segmentRefs.current[index] = el),
+        size,
+        type,
+      }),
+  )
+
+  return (
+    <div className={tabContainerStyles({ size, type })}>
+      <div className={activeTabStyles({ size, type })} style={indicatorStyle} />
+      {clonedChildren}
+    </div>
+  )
+}
+
+SegmentedControl.Root = SegmentedControl
+SegmentedControl.Button = Button
+SegmentedControl.IconButton = IconButton
+SegmentedControl.EmojiButton = EmojiButton
+
+export { SegmentedControl }
```

Comment:
This doesn't work in Server Components. Follow the same convention as in `<DropdownMenu />` or `<Tabs />`.